File: 프로그래머스/lv3/42898. 등굣길/등굣길.py
d=[[-1 for _ in range(101)] for _ in range(101)]
a=[[0 for _ in range(101)] for _ in range(101)]
def g(x,y,n,m):
    if x>=n or y>=m or a[x][y]:
        return 0
    if x==n-1 and y==m-1:
        return 1
    if d[x][y]!=-1:
        return d[x][y]
    ret=0
    ret+=(g(x+1,y,n,m)+g(x,y+1,n,m))%(int(1e9)+7)
    d[x][y]=ret
    return d[x][y]
def solution(m, n, puddles):
    for cur in puddles:
        a[cur[1]-1][cur[0]-1]=1
    return g(0,0,n,m)

File: 프로그래머스/lv3/42898. 등굣길/등굣길.java
import java.util.*;
class Solution {
    static int mod=1000000007;
    static int [][]dp=new int[101][101];
    static int N,M;
    static int [][]p=new int[101][101];
    public int go(int x,int y)
    {
        if (x>N || y>M || p[x][y]==1)return 0;
        if (x==N&&y==M)
            return 1;
        if (dp[x][y]!=-1)
            return dp[x][y];
        int ret=0;
        ret+=(go(x+1,y)+go(x,y+1))%mod;
        return dp[x][y]=ret;
    }
    public int solution(int m, int n, int[][] puddles) {
        int answer = 0;
        for (int i=0;i<=n;i++)
        {
            Arrays.fill(dp[i],-1);
        }
        for (int i=0;i<puddles.length;i++)
        {
            int x,y;
            x=puddles[i][1];
            y=puddles[i][0];
            p[x][y]=1;
        }
        N=n;
        M=m;
        return go(1,1);
    }
}